lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/ConsistentHashingCacheHostAddressProvider.java

```diff
@@ -0,0 +1,112 @@
+import jakarta.annotation.PreDestroy;
+import org.ishugaliy.allgood.consistent.hash.ConsistentHash;
+import org.ishugaliy.allgood.consistent.hash.HashRing;
+import org.ishugaliy.allgood.consistent.hash.hasher.DefaultHasher;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+
+public class ConsistentHashingCacheHostAddressProvider
+        implements CacheHostAddressProvider
+{
+    private static final Logger log = Logger.get(ConsistentHashingCacheHostAddressProvider.class);
+
+    private final NodeManager nodeManager;
+    private final ScheduledExecutorService hashRingUpdater = newSingleThreadScheduledExecutor(daemonThreadsNamed("hash-ring-refresher-%s"));
+
+    private final ConsistentHash<TrinoNode> consistentHashRing = HashRing.<TrinoNode>newBuilder()
+            .hasher(DefaultHasher.METRO_HASH)
+            .build();
+
+    @Inject
+    public ConsistentHashingCacheHostAddressProvider(NodeManager nodeManager)
+    {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+    }
+
+    @Override
+    public List<HostAddress> getHosts(String splitIdentifier)
+    {
+        return consistentHashRing.locate(splitIdentifier)
+                .stream()
+                .map(trinoNode -> trinoNode.node().getHostAndPort())
+                .collect(toImmutableList());
+    }
+
+    @PostConstruct
+    public void startRefreshingHashRing()
+    {
+        hashRingUpdater.scheduleWithFixedDelay(this::refreshHashRing, 5, 5, TimeUnit.SECONDS);
+        refreshHashRing();
+    }
+
+    @PreDestroy
+    public void destroy()
+    {
+        hashRingUpdater.shutdownNow();
+    }
+
+    @VisibleForTesting
+    synchronized void refreshHashRing()
+    {
+        try {
+            ImmutableSet<TrinoNode> trinoNodes = nodeManager.getWorkerNodes().stream().map(TrinoNode::new).collect(toImmutableSet());
+            Set<TrinoNode> hashRingNodes = consistentHashRing.getNodes();
+            Set<TrinoNode> removedNodes = Sets.difference(hashRingNodes, trinoNodes);
+            Set<TrinoNode> newNodes = Sets.difference(trinoNodes, hashRingNodes);
+            // Avoid acquiring a write lock in consistentHashRing if possible
+            if (!newNodes.isEmpty()) {
+                consistentHashRing.addAll(newNodes);
+            }
+            if (!removedNodes.isEmpty()) {
+                removedNodes.forEach(consistentHashRing::remove);
+            }
+        }
+        catch (Exception e) {
+            log.error(e, "Error refreshing hash ring");
+        }
+    }
+
+    private record TrinoNode(Node node)
```

Comment:
`private record TrinoNodeIdentifier(String nodeIdentifier)`

we don't need to retain full Node object here